Firestore/core/src/api/optional.h

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_API_OPTIONAL_H_
+#define FIRESTORE_CORE_SRC_API_OPTIONAL_H_
+
+#include <cassert>
+#include <cstdint>
+#include <new>
+#include <utility>
+
+#include "Firestore/core/src/api/common.h"
+
+namespace firebase {
+namespace firestore {
+namespace api {
+
+struct nullopt_t {
+  struct init {};
+  explicit nullopt_t(init) {
+  }
+};
+
+const nullopt_t nullopt((nullopt_t::init()));
+
+// optional<T> is a utility class to represent a value that can, but does not
+// have to be present.
+//
+// This has a few different use cases. For return values, it allows values to be
+// optionally returned without using pointers that might be null. In other
+// words, a function that returns a pointer might always return a valid pointer,
+// or it might sometimes return a valid pointer and sometimes return null.
+// Returning an optional makes it clear what the intention is.
+//
+// The other main use case is to be able to defer initialization of a value on
+// the stack or in a struct. By using an optional you can leave a value
+// uninitialized at creation and later initialize it when appropriate without
+// having to default construct a value, or keep a null pointer to a value and
+// later initializing it.
+template <typename T>
+class optional {
```

Comment:
Consider renaming this class to have an uppercase first letter (i.e. `Optional`). This would help differentiate it from the "optional" classes provided by absl and the stl and follows the naming convention of the other classes in the "api" directory.

Reply:
Done.